# MNIST Classification using fully connected Neural Networks
We will use keras a deep learning library with tensorflow as backend

1) First step is: the imports

In [1]:
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.optimizers import SGD
from keras.utils import to_categorical 

Using TensorFlow backend.


2) Second step is: data collection, 

#with the folloing instruction, we will download the mnist data from the server

In [2]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

#(x_train, y_train) is the training dataset, where (x_test, y_test) is the testing dataset

3) third step is: data preparation

#we will reshape the MNIST dataset to become 60,000 examples (ie. images) for the training dataset, and 10,000 examples (ie. images) for the testing dataet, and the input to the Network is 784 because each images has 28 x 28 pixels

In [3]:
x_train = x_train.reshape(60000, 784)
x_test  = x_test.reshape(10000, 784)
x_train = x_train.astype('float32')
x_test  = x_test.astype('float32')

4) fourth step is data normalization:

#Artificial neural networks provide good performance when the data is normalized

In [4]:
x_train /= 255
x_test /= 255

5) fifth step is data tranformation:

#we will convert the vector of labels (desired vector) to binary class matrices, thus each row in this matrix will become as one-hot encoding

In [5]:
y_train = to_categorical(y_train, 10)
y_test =  to_categorical(y_test, 10)

6) sixth step is: to define the model

#we will create one input layer that has 787 neurons, and two hidden layers, each layer has 128 neurons, finally an output layer that has 10 neurons

#RELU(x) = max(0,x) is a non-linear function
#softmax is another non-linear function, we use this function to squash the outputs of the Artificial neural networks to be between 0 and 1

In [6]:
model = Sequential()
model.add(Dense(128, activation='relu', input_shape=(784,)))
model.add(Dense(128, activation='relu'))
model.add(Dense(10, activation='softmax'))
model.summary()

7) seventh step is: to compile the created model, thus this model will become as a computaional graph

In [7]:
model.compile(loss='categorical_crossentropy',optimizer=SGD(), metrics=['accuracy'])

8) eighth step: after doing all above steps, we can start the model training 

#batch_size the number of training images in one forward/backward pass, this technique is faster and it has demonstrated that is better than using all dataset simultaneously

#epochs: one forward pass and one backward pass of all the training images

In [8]:
history = model.fit(x_train, y_train, batch_size=64, epochs=10, validation_data=(x_test, y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 6s 101us/step - loss: 0.8960 - acc: 0.7624 - val_loss: 0.4014 - val_acc: 0.8929
Epoch 2/10
60000/60000 [==============================] - 5s 80us/step - loss: 0.3647 - acc: 0.8981 - val_loss: 0.3126 - val_acc: 0.9103
Epoch 3/10
60000/60000 [==============================] - 5s 80us/step - loss: 0.3045 - acc: 0.9137 - val_loss: 0.2737 - val_acc: 0.9229
Epoch 4/10
60000/60000 [==============================] - 5s 80us/step - loss: 0.2709 - acc: 0.9228 - val_loss: 0.2460 - val_acc: 0.9292
Epoch 5/10
60000/60000 [==============================] - 5s 81us/step - loss: 0.2458 - acc: 0.9298 - val_loss: 0.2283 - val_acc: 0.9339
Epoch 6/10
60000/60000 [==============================] - 5s 80us/step - loss: 0.2256 - acc: 0.9357 - val_loss: 0.2115 - val_acc: 0.9377
Epoch 7/10
60000/60000 [==============================] - 5s 79us/step - loss: 0.2087 - acc: 0.9408 - val_loss: 0.1986 - val_ac

9) the last step is model evaluation

In [9]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.16808781869783998
Test accuracy: 0.9509


#After developing fully connected Artifial Neural Networks, we have obtained a test accuracy = 0.95, and a test loss = 0.16

# To improve the proposed fully connected model, there are many methods and steps

1) first step: adding more epochs, for examples epochs = 30

In [10]:
history = model.fit(x_train, y_train, batch_size=64, epochs=30, validation_data=(x_test, y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/30
60000/60000 [==============================] - 5s 87us/step - loss: 0.1622 - acc: 0.9540 - val_loss: 0.1585 - val_acc: 0.9534
Epoch 2/30
60000/60000 [==============================] - 5s 82us/step - loss: 0.1535 - acc: 0.9562 - val_loss: 0.1532 - val_acc: 0.9550
Epoch 3/30
60000/60000 [==============================] - 5s 82us/step - loss: 0.1460 - acc: 0.9591 - val_loss: 0.1466 - val_acc: 0.9565
Epoch 4/30
60000/60000 [==============================] - 6s 94us/step - loss: 0.1389 - acc: 0.9609 - val_loss: 0.1413 - val_acc: 0.9582
Epoch 5/30
60000/60000 [==============================] - 5s 89us/step - loss: 0.1326 - acc: 0.9631 - val_loss: 0.1361 - val_acc: 0.9602
Epoch 6/30
60000/60000 [==============================] - 5s 87us/step - loss: 0.1266 - acc: 0.9643 - val_loss: 0.1304 - val_acc: 0.9614
Epoch 7/30
60000/60000 [==============================] - 5s 88us/step - loss: 0.1215 - acc: 0.9663 - val_loss: 0.1262 - val_acc

In [ ]:
#with only epochs =30, we have obtianed 

In [11]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.08447478631325066
Test accuracy: 0.974


#with only epochs = 30, we have obtained a test accuracy = 0.974, and a test loss = 0.084, these results with adding more epochs we have obtained better results than the above results

2) second step: to obtain more accurate results, we adjust the batch size, for examples batch size = 128

In [12]:
history = model.fit(x_train, y_train, batch_size=64, epochs=30, validation_data=(x_test, y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/30
60000/60000 [==============================] - 5s 85us/step - loss: 0.0540 - acc: 0.9855 - val_loss: 0.0845 - val_acc: 0.9741
Epoch 2/30
60000/60000 [==============================] - 5s 80us/step - loss: 0.0525 - acc: 0.9857 - val_loss: 0.0848 - val_acc: 0.9736
Epoch 3/30
60000/60000 [==============================] - 5s 80us/step - loss: 0.0514 - acc: 0.9861 - val_loss: 0.0839 - val_acc: 0.9743
Epoch 4/30
60000/60000 [==============================] - 5s 82us/step - loss: 0.0497 - acc: 0.9869 - val_loss: 0.0817 - val_acc: 0.9746
Epoch 5/30
60000/60000 [==============================] - 5s 81us/step - loss: 0.0484 - acc: 0.9871 - val_loss: 0.0812 - val_acc: 0.9746
Epoch 6/30
60000/60000 [==============================] - 5s 82us/step - loss: 0.0469 - acc: 0.9877 - val_loss: 0.0809 - val_acc: 0.9748
Epoch 7/30
60000/60000 [==============================] - 5s 81us/step - loss: 0.0459 - acc: 0.9879 - val_loss: 0.0818 - val_acc

In [13]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.07582007238190853
Test accuracy: 0.9769


#with batch size = 128, we have obtained a test accuracy = 0.976, and a test loss = 0.075, these results is betten than the above results

3) to avoid the problem of unbalanced data, we add a weighted cross entropy loss function instead of cross entropy loss function
there are several methods to get the best weight for each class, but here we will assign each weight to 0.1

In [15]:
classes= [0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1]
history = model.fit(x_train, y_train, batch_size=128, epochs=30, validation_data=(x_test, y_test), class_weight=classes)

Train on 60000 samples, validate on 10000 samples
Epoch 1/30
60000/60000 [==============================] - 3s 47us/step - loss: 0.0241 - acc: 0.9952 - val_loss: 0.0737 - val_acc: 0.9770
Epoch 2/30
60000/60000 [==============================] - 3s 45us/step - loss: 0.0237 - acc: 0.9953 - val_loss: 0.0743 - val_acc: 0.9766
Epoch 3/30
60000/60000 [==============================] - 3s 45us/step - loss: 0.0235 - acc: 0.9952 - val_loss: 0.0754 - val_acc: 0.9768
Epoch 4/30
60000/60000 [==============================] - 3s 45us/step - loss: 0.0231 - acc: 0.9953 - val_loss: 0.0745 - val_acc: 0.9765
Epoch 5/30
60000/60000 [==============================] - 3s 46us/step - loss: 0.0229 - acc: 0.9956 - val_loss: 0.0734 - val_acc: 0.9770
Epoch 6/30
60000/60000 [==============================] - 3s 46us/step - loss: 0.0226 - acc: 0.9956 - val_loss: 0.0734 - val_acc: 0.9771
Epoch 7/30
60000/60000 [==============================] - 3s 45us/step - loss: 0.0223 - acc: 0.9957 - val_loss: 0.0739 - val_acc

In [16]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.07325211380326654
Test accuracy: 0.9777


#with classes= [0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1], we have obtained a test accuracy = 0.977, and a test loss = 0.073, also these results is betten than the above results 

4) the fourth step: changing the architecture, we will add more neurons in each hidden layer

In [17]:
model = Sequential()
model.add(Dense(256, activation='relu', input_shape=(784,)))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [18]:
model.compile(loss='categorical_crossentropy',optimizer=SGD(), metrics=['accuracy'])

In [23]:
classes= [0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1]
history = model.fit(x_train, y_train, batch_size=40, epochs=20, validation_data=(x_test, y_test), class_weight=classes)

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 9s 150us/step - loss: 0.0559 - acc: 0.9847 - val_loss: 0.0815 - val_acc: 0.9772
Epoch 2/20
60000/60000 [==============================] - 9s 147us/step - loss: 0.0532 - acc: 0.9850 - val_loss: 0.0783 - val_acc: 0.9752
Epoch 3/20
60000/60000 [==============================] - 9s 146us/step - loss: 0.0507 - acc: 0.9864 - val_loss: 0.0785 - val_acc: 0.9763
Epoch 4/20
60000/60000 [==============================] - 9s 146us/step - loss: 0.0483 - acc: 0.9870 - val_loss: 0.0783 - val_acc: 0.9763
Epoch 5/20
60000/60000 [==============================] - 9s 148us/step - loss: 0.0460 - acc: 0.9877 - val_loss: 0.0762 - val_acc: 0.9774
Epoch 6/20
60000/60000 [==============================] - 9s 146us/step - loss: 0.0439 - acc: 0.9885 - val_loss: 0.0763 - val_acc: 0.9773
Epoch 7/20
60000/60000 [==============================] - 9s 148us/step - loss: 0.0419 - acc: 0.9892 - val_loss: 0.0739 - 

In [24]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.06823914499669335
Test accuracy: 0.9794


#with these changes of the Neural architecture, we have obtained a test accuracy = 0.979, and a test loss = 0.068, in which these results is better than the above results

5) the last step: we will add more layers, 3 hidden layers, each with 256 neurons

In [25]:
model = Sequential()
model.add(Dense(256, activation='relu', input_shape=(784,)))
model.add(Dense(256, activation='relu'))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [26]:
model.compile(loss='categorical_crossentropy',optimizer=SGD(), metrics=['accuracy'])

In [31]:
history = model.fit(x_train, y_train, batch_size=32, epochs=30, validation_data=(x_test, y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/30
60000/60000 [==============================] - 12s 198us/step - loss: 0.0078 - acc: 0.9992 - val_loss: 0.0764 - val_acc: 0.9796
Epoch 2/30
60000/60000 [==============================] - 12s 196us/step - loss: 0.0071 - acc: 0.9994 - val_loss: 0.0740 - val_acc: 0.9795
Epoch 3/30
60000/60000 [==============================] - 12s 196us/step - loss: 0.0066 - acc: 0.9995 - val_loss: 0.0756 - val_acc: 0.9789
Epoch 4/30
60000/60000 [==============================] - 12s 196us/step - loss: 0.0059 - acc: 0.9995 - val_loss: 0.0734 - val_acc: 0.9802
Epoch 5/30
60000/60000 [==============================] - 12s 201us/step - loss: 0.0055 - acc: 0.9996 - val_loss: 0.0738 - val_acc: 0.9804
Epoch 6/30
60000/60000 [==============================] - 12s 202us/step - loss: 0.0052 - acc: 0.9997 - val_loss: 0.0776 - val_acc: 0.9803
Epoch 7/30
60000/60000 [==============================] - 12s 199us/step - loss: 0.0046 - acc: 0.9997 - val_loss: 0.

In [32]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.08305254170282296
Test accuracy: 0.9806


#we have added only one layer, and we have obtained a test accuracy = 0.98, and a test loss = 0.083.

In the next part, we will use Convolutional Neural Networks to improve the results on the testing dataset
